### PR TITLE
Citation System

### PR DESCRIPTION
**Pull request by Royal-lobster**
Closes https://github.com/EveripediaNetwork/issues/issues/293
Fixes https://github.com/EveripediaNetwork/issues/issues/306
Fixes https://github.com/EveripediaNetwork/issues/issues/319

# Changes
1. Implemented cite plugin in editor to add citations
2. Added new references metadata to wikis for storing citation data
3. Added references section in wiki and revision pages and make cite markers to link references from this section
4. Added link previews for cite markers in wiki and revision pages
5. Added ability to remove, edit, cite from and search existing references within citation plugin
7. ⚠️ Removed ```md``` and ```setMd``` in ```create wiki``` and made the editor content to store directly to ```wiki slice``` to make updates to editor outside create-wiki (It is working fine but might required some more testing)
8. Fixed cite marks and toc reduplicating in some cases, hence refactored ```addToToc``` to a seperate file

# Links
- https://monopedia-ui-git-citations-prediqt.vercel.app/wiki/wiki-citation-test
- https://monopedia-ui-git-citations-prediqt.vercel.app/create-wiki?slug=wiki-citation-test

# Screenshots
Frame 171

# Tasks
- [x] Create plugin files and icons and connect to editor
- [x] Implement ui for editor plugin for cite input
- [x] Implement cite symbols in wysiwyg 
- [x] Add citations to wiki metadata slice
- [x] Show already linked references in plugin frame
- [x] Create a section for citations in wiki page
- [x] Make a custom renderer for showing citation info in wiki page on hover
- [x] decide on markdown syntax for citation 
- [x] Provide option to edit/delete references in create-wiki
- [x] Add search to existing citations
- [x] Make backward links from reference box to all cite marks for that reference
- [x] Fix cite marks reduplicating in some cases

### CODE REVIEW

**Comment by kesar** [comment]
as part of this PR pls remove the linking plugin in the editor as well 🙏🏻

File: src/editor-plugins/cite/frame/CiteFrame.tsx [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+import React, { useEffect } from 'react'
+
+const FrameTab = ({ children }: { children: React.ReactNode }) => (
+  <Tab
+    border="unset"
+    h="unset !important"
+    borderBottom="1px solid #dbdde5 !important"
+    borderRadius="0 !important"
+    sx={{
+      '&[aria-selected="true"]': {
```
Comment (anubra266) [inline src/editor-plugins/cite/frame/CiteFrame.tsx:28]: There's a `_selected` prop. https://chakra-ui.com/docs/components/disclosure/tabs#styling-the-tab-states-via-props

File: src/editor-plugins/cite/frame/CiteFrame.tsx [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        borderBottom: '2px solid #4ba6f8 !important',
+        color: '#4ba6f8 !important',
+      },
+    }}
+    _dark={{
+      color: 'white !important',
+      borderBottom: '1px solid #2b2f37 !important',
+      '&[aria-selected="true"]': {
```
Comment (anubra266) [inline src/editor-plugins/cite/frame/CiteFrame.tsx:36]: You can also do something like
```js
_selected={{
   ...,
   _dark: {...}
}}
```

**Comment by anubra266** [comment]
@Royal-lobster Why is `!important` needed more majority of the styling props?

File: src/editor-plugins/cite/frame/CiteFrame.tsx (hunk 2 above)
Comment (Royal-lobster) [reply src/editor-plugins/cite/frame/CiteFrame.tsx:36]: Ohh that's awesome Thanks !

File: src/editor-plugins/cite/frame/index.tsx [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+      type: 'wiki/updateMetadata',
+      payload: {
+        id: 'references',
+        value: JSON.stringify([
+          ...references,
+          {
+            id: newRefId,
+            url,
+            description,
+            timestamp: new Date().toISOString(),
```
Comment (kesar) [inline src/editor-plugins/cite/frame/index.tsx:70]: perhaps we could use unixtime so its more universal and then we can default to convert it in the UI to whatever we want

File: src/components/Wiki/WikiPage/CitationPreviewHover.tsx [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+      },
+    )
+    wiki = revisionData?.content[0]
+  }
+
+  if (!wiki)
+    return (
+      <Link href={href}>
+        <sup>{text}</sup>
+      </Link>
```
Comment (anubra266) [inline src/components/Wiki/WikiPage/CitationPreviewHover.tsx:58]: Could just do
```suggestion
      <Link as="sup" href={href}>{text}</Link>

File: src/components/Wiki/WikiPage/CitationPreviewHover.tsx [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        >
+          <PopoverArrow />
+          <PopoverBody>
+            <HStack
+              pb={2}
+              mb={2}
+              borderBottomWidth="1px"
+              justify="space-between"
+            >
+              <Tag colorScheme="blue" size="sm" fontWeight="500">
```
Comment (anubra266) [inline src/components/Wiki/WikiPage/CitationPreviewHover.tsx:102]: It's preferrable to use tokens, so any general theme shift could be done with ease.
```suggestion
              <Tag colorScheme="blue" size="sm" fontWeight="medium">

File: src/components/Wiki/WikiPage/CitationPreviewHover.tsx [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+                <Link
+                  color="#304262 !important"
+                  _dark={{
+                    color: '#9dcbf0 !important',
+                  }}
+                  fontSize="sm"
+                  p="0 0 0 0 !important"
```
Comment (anubra266) [inline src/components/Wiki/WikiPage/CitationPreviewHover.tsx:109]: could just do:
```suggestion
                  p="0 !important"

File: src/components/Wiki/WikiPage/WikiReferences.tsx [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    window.addEventListener('hashchange', onHashChanged)
+    return () => {
+      window.removeEventListener('hashchange', onHashChanged)
+    }
+  }, [])
+
+  if (references.length === 0) return null
+  return (
+    <Box borderTopWidth="1px" p={4}>
+      <Heading my={4} p={2} fontWeight="500">
```
Comment (anubra266) [inline src/components/Wiki/WikiPage/WikiReferences.tsx:31]:
```suggestion
      <Heading my={4} p={2} fontWeight="medium">

File: src/components/Wiki/WikiPage/WikiReferences.tsx [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        REFERENCES
+      </Heading>
+      <SimpleGrid mb={8} columns={[1, 2, 3]} spacing={4}>
+        {references.map((ref, index) => (
+          <Flex
+            id={`cite-id-${ref.id}`}
+            bgColor={
```
Comment (anubra266) [inline src/components/Wiki/WikiPage/WikiReferences.tsx:38]: This is okay, but just thought you should know that `bg` is also okay here. doesn't have to be `bgColor`
```suggestion
            bg={

**Review by kesar — APPROVED** [review]
great work 👍🏻

**Pull request merged**